File: agent/agent.py
#!/usr/bin/env python3
"""
Tunnel Agent — connects your local server to a public URL.

Usage:
    python agent.py --alias my-app --token tk_... --port 3000

Config file (~/.tunnelrc):
    The agent can save your settings so you don't need to type them every time.
    On first run with --save, it writes to ~/.tunnelrc:
      [my-app]
      token = tk_...
      port = 3000

Options:
    --alias     Tunnel alias (required)
    --token     Secret tunnel token (required if not in config)
    --port      Local port to forward to (default: 3000)
    --host      Local host (default: localhost)
    --server    Backend WebSocket URL (default: ws://localhost:8000)
    --save      Save alias/token/port to ~/.tunnelrc for future use
    --force     Not implemented yet (reserved for future use)
"""

import argparse
import asyncio
import base64
import configparser
import json
import os
import signal
import sys
import time
import uuid
from datetime import datetime
from pathlib import Path

try:
    import requests
    import websockets
    from websockets.exceptions import ConnectionClosedError, InvalidURI, WebSocketException
except ImportError:
    print("\n" + "!" * 50)
    print("❌ ERROR: Missing dependencies!")
    print("Please run: pip install websockets requests")
    print("!" * 50 + "\n")
    sys.exit(1)

# ─── Config ────────────────────────────────────────────────────────────────

CONFIG_PATH = Path.home() / ".tunnelrc"

# Reconnect strategy
MAX_RETRIES = 10
BACKOFF_BASE = 1      # seconds
BACKOFF_MAX = 30      # seconds cap

# ─── Colors ────────────────────────────────────────────────────────────────

GREEN  = "\033[92m"
RED    = "\033[91m"
YELLOW = "\033[93m"
CYAN   = "\033[96m"
GRAY   = "\033[90m"
BOLD   = "\033[1m"
RESET  = "\033[0m"

STATUS_200 = GREEN
STATUS_4XX = YELLOW
STATUS_5XX = RED


def color_status(status: int) -> str:
    if 200 <= status < 300:
        return GREEN + str(status) + RESET
    elif 400 <= status < 500:
        return YELLOW + str(status) + RESET
    else:
        return RED + str(status) + RESET


# ─── Banner ────────────────────────────────────────────────────────────────

def print_banner(alias: str, public_url: str, local_url: str):
    print()
    print(f"{BOLD}{CYAN}  ╔══════════════════════════════════════════════╗{RESET}")
    print(f"{BOLD}{CYAN}  ║          TUNNEL AGENT  v1.0.0               ║{RESET}")
    print(f"{BOLD}{CYAN}  ╚══════════════════════════════════════════════╝{RESET}")
    print()
    print(f"  {BOLD}Alias   {RESET}: {CYAN}{alias}{RESET}")
    print(f"  {BOLD}Public  {RESET}: {GREEN}{public_url}{RESET}")
    print(f"  {BOLD}Local   {RESET}: {GRAY}{local_url}{RESET}")
    print()
    print(f"  {GRAY}Connecting to server...{RESET}")


def print_connected():
    print(f"  {GREEN}✅ Connected! Tunnel is live.{RESET}")
    print()
    print(f"  {GRAY}{'Timestamp':<14} {'Method':<8} {'Path':<30} {'Status':<8} {'Latency'}{RESET}")
    print(f"  {GRAY}{'─'*14} {'─'*8} {'─'*30} {'─'*8} {'─'*8}{RESET}")


def print_request_log(method: str, path: str, status: int, latency_ms: float):
    ts = datetime.now().strftime("%H:%M:%S")
    path_display = path[:29] if len(path) > 29 else path
    print(f"  {GRAY}{ts:<14}{RESET} {CYAN}{method:<8}{RESET} {path_display:<30} {color_status(status):<24} {GRAY}{latency_ms:.0f}ms{RESET}")


# ─── Config File Helpers ────────────────────────────────────────────────────

def load_config(alias: str) -> dict:
    """Load token and port from ~/.tunnelrc for a given alias."""
    config = configparser.ConfigParser()
    if CONFIG_PATH.exists():
        config.read(CONFIG_PATH)
        if alias in config:
            return dict(config[alias])
    return {}


def save_config(alias: str, token: str, port: int):
    """Save alias config to ~/.tunnelrc."""
    config = configparser.ConfigParser()
    if CONFIG_PATH.exists():
        config.read(CONFIG_PATH)
    if alias not in config:
        config[alias] = {}
    config[alias]["token"] = token
    config[alias]["port"] = str(port)
    with open(CONFIG_PATH, "w") as f:
        config.write(f)
    print(f"  {GREEN}✅ Config saved to {CONFIG_PATH}{RESET}")


# ─── Request Forwarding ────────────────────────────────────────────────────

def forward_to_local(method: str, path: str, headers: dict, body: bytes, local_url: str) -> dict:
    """
    Forward an HTTP request to the local server and return a response dict.
    """
    # Build clean headers for local request
    forward_headers = {}
    for key, val in headers.items():
        # Strip Django META prefix
        if key.startswith("HTTP_"):
            clean = key[5:].replace("_", "-").title()
            forward_headers[clean] = val
        elif key == "CONTENT_TYPE":
            forward_headers["Content-Type"] = val
        elif key == "CONTENT_LENGTH" and val:
            forward_headers["Content-Length"] = val

    # Remove headers that shouldn't be forwarded
    forward_headers.pop("Host", None)

    url = local_url.rstrip("/") + path.split("?")[0]
    params = path.split("?")[1] if "?" in path else None

    try:
        start = time.time()
        resp = requests.request(
            method=method,
            url=url,
            headers=forward_headers,
            data=body,
            params=params,
            timeout=25,
            allow_redirects=True,
        )
        latency_ms = (time.time() - start) * 1000

        # Strip headers that we shouldn't pass back (compression, chunks, etc)
        # The agent de-chunks and de-compresses automatically via 'requests'
        response_headers = dict(resp.headers)
        response_headers.pop("Content-Encoding", None)
        response_headers.pop("Transfer-Encoding", None)
        response_headers.pop("Content-Length", None)

        return {
            "status": resp.status_code,
            "headers": response_headers,
            "body": base64.b64encode(resp.content).decode("utf-8"),
            "latency_ms": latency_ms,
        }

    except requests.exceptions.ConnectionError:
        error_msg = f"Local server at {local_url} is not responding. Is your app running?"
        return {
            "status": 503,
            "headers": {"Content-Type": "text/plain"},
            "body": base64.b64encode(error_msg.encode()).decode(),
            "latency_ms": 0,
        }
    except requests.exceptions.Timeout:
        error_msg = "Local server timed out (25s)."
        return {
            "status": 504,
            "headers": {"Content-Type": "text/plain"},
            "body": base64.b64encode(error_msg.encode()).decode(),
            "latency_ms": 25000,
        }
    except Exception as e:
        error_msg = f"Agent error: {e}"
        return {
            "status": 502,
            "headers": {"Content-Type": "text/plain"},
            "body": base64.b64encode(error_msg.encode()).decode(),
            "latency_ms": 0,
        }


# ─── WebSocket Connection ───────────────────────────────────────────────────

async def run_agent(alias: str, token: str, local_url: str, ws_url: str):
    """
    Main agent loop: connect, handle requests, auto-reconnect on drop.
    """
    retry_count = 0

    while True:
        try:
            async with websockets.connect(
                ws_url,
                additional_headers={"Authorization": f"Bearer {token}"},
                ping_interval=20,
                ping_timeout=30,
            ) as ws:
                retry_count = 0
                print_connected()

                async for raw_message in ws:
                    try:
                        msg = json.loads(raw_message)
                    except json.JSONDecodeError:
                        continue

                    msg_type = msg.get("type")

                    # Server sent a disconnect reason (e.g. bad token, alias taken)
                    if msg_type == "disconnect":
                        reason = msg.get("reason", "Unknown reason.")
                        print(f"\n  {RED}❌ {reason}{RESET}")
                        sys.exit(1)

                    # Server is forwarding an HTTP request to us
                    elif msg_type == "request":
                        request_id = msg.get("request_id")
                        method = msg.get("method", "GET")
                        path = msg.get("path", "/")
                        headers = msg.get("headers", {})
                        body = msg.get("body", "")

                        # Decode incoming base64 body from server
                        try:
                            decoded_body = base64.b64decode(body)
                        except Exception:
                            decoded_body = b""

                        # Forward to local server
                        result = forward_to_local(method, path, headers, decoded_body, local_url)

                        # Log it
                        print_request_log(method, path, result["status"], result["latency_ms"])

                        # Send response back to server
                        await ws.send(json.dumps({
                            "type": "response",
                            "request_id": request_id,
                            "status": result["status"],
                            "headers": result["headers"],
                            "body": result["body"],
                        }))

        except ConnectionClosedError as e:
            print(f"\n  {YELLOW}⚠️  Connection dropped (code {e.code}).{RESET}")

        except (OSError, WebSocketException) as e:
            print(f"\n  {YELLOW}⚠️  Connection error: {e}{RESET}")

        except KeyboardInterrupt:
            print(f"\n\n  {CYAN}👋 Gracefully disconnecting...{RESET}\n")
            sys.exit(0)

        # Reconnect logic
        retry_count += 1
        if retry_count > MAX_RETRIES:
            print(f"\n  {RED}❌ Could not reconnect after {MAX_RETRIES} attempts. Run the agent again manually.{RESET}\n")
            sys.exit(1)

        wait = min(BACKOFF_BASE * (2 ** (retry_count - 1)), BACKOFF_MAX)
        print(f"  {YELLOW}↻  Retrying in {wait:.0f}s... (attempt {retry_count}/{MAX_RETRIES}){RESET}")
        await asyncio.sleep(wait)


# ─── Entry Point ───────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description="Tunnel Agent — expose your local server via a public URL."
    )
    parser.add_argument("--alias",  type=str, help="Tunnel alias (e.g. my-app)")
    parser.add_argument("--token",  type=str, help="Secret tunnel token")
    parser.add_argument("--port",   type=int, help="Local port to forward to (default: 3000)")
    parser.add_argument("--host",   type=str, default="localhost", help="Local host (default: localhost)")
    parser.add_argument("--server", type=str, default="ws://localhost:8000", help="Backend WS URL")
    parser.add_argument("--save",   action="store_true", help="Save config to ~/.tunnelrc")
    args = parser.parse_args()

    alias = args.alias
    token = args.token
    port  = args.port

    # Try to load from config file if alias given but no token
    if alias and not token:
        saved = load_config(alias)
        token = saved.get("token")
        if not port:
            port = int(saved.get("port", 3000))

    port = port or 3000

    # Validate required args
    if not alias:
        print(f"{RED}❌ --alias is required.{RESET}")
        parser.print_help()
        sys.exit(1)
    if not token:
        print(f"{RED}❌ --token is required (or save it in ~/.tunnelrc with --save).{RESET}")
        parser.print_help()
        sys.exit(1)

    # Optionally save to config
    if args.save:
        save_config(alias, token, port)

    local_url = f"http://{args.host}:{port}"
    ws_url    = f"{args.server.rstrip('/')}/ws/tunnel/{alias}/"
    # Build a rough public URL for display purposes
    public_url = args.server.replace("ws://", "http://").replace("wss://", "https://") + f"/t/{alias}/"

    print_banner(alias, public_url, local_url)

    # Run
    try:
        asyncio.run(run_agent(alias, token, local_url, ws_url))
    except KeyboardInterrupt:
        print(f"\n  {CYAN}👋 Goodbye!{RESET}\n")


if __name__ == "__main__":
    main()

File: agent/agent.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const http = require('http');
const { execSync } = require('child_process');

// ─── Colors ────────────────────────────────────────────────────────────────
const GREEN = "\x1b[38;5;82m";
const RED = "\x1b[38;5;196m";
const YELLOW = "\x1b[38;5;214m";
const CYAN = "\x1b[38;5;39m";
const GRAY = "\x1b[38;5;244m";
const WHITE = "\x1b[38;5;255m";
const BOLD = "\x1b[1m";
const RESET = "\x1b[0m";

function colorStatus(status) {
    if (status >= 200 && status < 300) return GREEN + status + RESET;
    if (status >= 400 && status < 500) return YELLOW + status + RESET;
    return RED + status + RESET;
}

// ─── Dependency Check ──────────────────────────────────────────────────────
function ensureDependencies() {
    const missing = [];
    try { require.resolve('ws'); } catch (e) { missing.push('ws'); }
    try { require.resolve('axios'); } catch (e) { missing.push('axios'); }
    try { require.resolve('minimist'); } catch (e) { missing.push('minimist'); }

    if (missing.length === 0) return;

    console.log(`\n${BOLD}${YELLOW}  ⚠ Missing required dependencies: ${missing.join(', ')}${RESET}`);
    console.log(`  ${GRAY}The agent needs these to establish a secure tunnel.${RESET}`);

    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return new Promise((resolve) => {
        readline.question(`\n  ${BOLD}Apply missing patches via npm? [Y/n]: ${RESET}`, (answer) => {
            readline.close();
            const choice = answer.trim().toLowerCase();
            if (choice === "" || choice === "y" || choice === "yes") {
                console.log(`  ${CYAN}# Downloading patches...${RESET}`);
                try {
                    execSync(`npm install ${missing.join(' ')}`, { stdio: 'ignore' });
                    console.log(`  ${GREEN}✓ Patches applied successfully.${RESET}`);
                    resolve();
                } catch (err) {
                    console.log(`\n${BOLD}${RED}  ❌ Critical Failure: Could not install dependencies.${RESET}`);
                    console.log(`  ${GRAY}${err.message}${RESET}`);
                    console.log(`\n  ${BOLD}Please run manually:${RESET}`);
                    console.log(`  ${WHITE}npm install ${missing.join(' ')}${RESET}\n`);
                    process.exit(1);
                }
            } else {
                console.log(`\n  ${RED}❌ Agent cannot proceed without dependencies.${RESET}\n`);
                process.exit(1);
            }
        });
    });
}

(async () => {
    await ensureDependencies();

    const axios = require('axios');
    const WebSocket = require('ws');

    // ─── Arguments ─────────────────────────────────────────────────────────────
    const args = require('minimist')(process.argv.slice(2), {
        string: ['alias', 'token', 'host', 'server'],
        number: ['port'],
        boolean: ['save', 'verbose'],
        default: {
            port: 3000,
            host: 'localhost',
            server: 'ws://localhost:8000',
            verbose: false
        }
    });

    if (!args.alias || !args.token) {
        console.log(`${RED}❌ --alias and --token are required.${RESET}`);
        console.log(`Example: node agent.js --alias my-app --token tk_...`);
        process.exit(1);
    }

    const { alias, token, port, host, server, verbose } = args;
    const localUrl = `http://${host}:${port}`;
    const wsUrl = `${server.replace(/\/$/, '')}/ws/tunnel/${alias}/`;
    const publicUrl = server.replace('ws://', 'http://').replace('wss://', 'https://') + `/t/${alias}/`;

    // ─── Banner ────────────────────────────────────────────────────────────
    console.log(`\n${BOLD}${CYAN}  ┌────────────────────────────────────────────────────────────┐${RESET}`);
    console.log(`${BOLD}${CYAN}  │${RESET}  {BOLD}MATRIX TUNNEL AGENT (NODE) v1.0.0{RESET}                   ${BOLD}${CYAN}│${RESET}`);
    console.log(`${BOLD}${CYAN}  └────────────────────────────────────────────────────────────┘${RESET}\n`);

    console.log(`  ${BOLD}${GRAY}NODE ALIAS${RESET}   : ${BOLD}${WHITE}${alias}${RESET}`);
    console.log(`  ${BOLD}${GRAY}PUBLIC ENDPT${RESET} : ${CYAN}${publicUrl}${RESET}`);
    console.log(`  ${BOLD}${GRAY}LOCAL BRIDGE${RESET} : ${GRAY}${localUrl}${RESET}`);
    console.log(`\n  ${GRAY}# Initializing secure uplink...${RESET}`);

    function printConnected() {
        console.log(`  ${GREEN}✓ Connection Established. Node is live.${RESET}\n`);
        const header = `  ${GRAY}${'TIMESTAMP'.padEnd(12)} ${'METHOD'.padEnd(10)} ${'PATH'.padEnd(32)} ${'STATUS'.padEnd(12)} ${'LATENCY'.padEnd(10)}${RESET}`;
        const divider = `  ${GRAY}${'─'.repeat(10).padEnd(12)} ${'─'.repeat(8).padEnd(10)} ${'─'.repeat(30).padEnd(32)} ${'─'.repeat(10).padEnd(12)} ${'─'.repeat(8).padEnd(10)}${RESET}`;
        console.log(header);
        console.log(divider);
    }

    function shouldLog(path, verbose) {
        if (verbose) return true;
        // Aggressive Noise filters
        const noisePrefixes = ["/_next/", "/static/", "/media/", "/favicon.ico", "/__nextjs_launcher"];
        const noiseExtensions = [".png", ".jpg", ".jpeg", ".gif", ".svg", ".css", ".map", ".ico", ".js", ".json"];

        if (noisePrefixes.some(p => path.startsWith(p))) return false;
        if (noiseExtensions.some(e => path.toLowerCase().endsWith(e))) return false;
        return true;
    }

    function printRequestLog(method, path, status, latencyMs) {
        if (!shouldLog(path, verbose)) return;

        const ts = new Date().toLocaleTimeString('en-GB', { hour12: false });
        const pathDisplay = path.length > 30 ? path.substring(0, 29) + "…" : path;

        let methodCol = CYAN;
        if (['POST', 'PUT'].includes(method)) methodCol = YELLOW;
        else if (method === 'DELETE') methodCol = RED;

        const log = `  ${GRAY}${ts.padEnd(12)}${RESET} ${methodCol}${BOLD}${method.padEnd(10)}${RESET} ${pathDisplay.padEnd(32)} ${colorStatus(status).padEnd(22)} ${GRAY}${Math.round(latencyMs).toString().padStart(6)}ms${RESET}`;
        console.log(log);
    }

    // ─── Logic ─────────────────────────────────────────────────────────────
    const connect = () => {
        const ws = new WebSocket(wsUrl, {
            headers: { 'Authorization': `Bearer ${token}` }
        });

        ws.on('open', () => {
            printConnected();
        });

        ws.on('message', async (data) => {
            let msg;
            try { msg = JSON.parse(data); } catch (e) { return; }

            if (msg.type === 'disconnect') {
                console.log(`\n  ${RED}❌ ${msg.reason || 'Unknown reason.'}${RESET}`);
                process.exit(1);
            }

            if (msg.type === 'request') {
                const start = Date.now();
                const { request_id, method, path, headers, body } = msg;

                try {
                    const response = await axios({
                        method,
                        url: localUrl.replace(/\/$/, '') + path,
                        headers: Object.fromEntries(
                            Object.entries(headers).map(([k, v]) => {
                                if (k.startsWith('HTTP_')) return [k.substring(5).replace(/_/g, '-').toLowerCase(), v];
                                if (k === 'CONTENT_TYPE') return ['content-type', v];
                                return [k.toLowerCase(), v];
                            }).filter(([k]) => !['host'].includes(k))
                        ),
                        data: body ? Buffer.from(body, 'base64') : undefined,
                        maxRedirects: 5,
                        validateStatus: () => true,
                        responseType: 'arraybuffer'
                    });

                    const latencyMs = Date.now() - start;
                    printRequestLog(method, path, response.status, latencyMs);

                    ws.send(JSON.stringify({
                        type: 'response',
                        request_id,
                        status: response.status,
                        headers: response.headers,
                        body: Buffer.from(response.data).toString('base64')
                    }));
                } catch (err) {
                    const status = err.response ? err.response.status : 503;
                    const errorMsg = err.response ? err.message : `Local server at ${localUrl} is not responding.`;
                    printRequestLog(method, path, status, 0);

                    ws.send(JSON.stringify({
                        type: 'response',
                        request_id,
                        status,
                        headers: { 'content-type': 'text/plain' },
                        body: Buffer.from(errorMsg).toString('base64')
                    }));
                }
            }
        });

        ws.on('error', (err) => {
            console.log(`\n  ${YELLOW}⚠ Warning: Network error detected: ${err.message}${RESET}`);
        });

        ws.on('close', () => {
            console.log(`\n  ${YELLOW}⚠ Warning: Uplink interrupted. Retrying...${RESET}`);
            setTimeout(connect, 5000);
        });
    };

    connect();

    process.on('SIGINT', () => {
        console.log(`\n\n  ${CYAN}⠿ Gracefully severing neural uplink...${RESET}\n`);
        process.exit(0);
    });

})();

function padRight(str, len) {
    return str + ' '.repeat(Math.max(0, len - str.length));
}
